clients/src/main/java/org/apache/kafka/common/protocol/ApiVersion.java

```diff
@@ -53,4 +53,18 @@ public String toString() {
             ", maxVersion= " + maxVersion +
             ")";
     }
+
+    public static ApiVersion versionsInCommon(ApiKeys apiKey, ApiVersion supportedVersions,
+                                              short minAllowedVersion, short maxAllowedVersion) {
+        if (supportedVersions == null)
+            throw new UnsupportedVersionException("The broker does not support " + apiKey);
+
+        short minVersion = (short) Math.max(minAllowedVersion, supportedVersions.minVersion);
+        short maxVersion = (short) Math.min(maxAllowedVersion, supportedVersions.maxVersion);
+        if (minVersion > maxVersion)
+            throw new UnsupportedVersionException("The broker does not support " + apiKey +
```

Comment:
Not totally sure this gets thrown in the right place. If there are no overlapping versions, perhaps we should leave the api out of the result. Perhaps that suggests we should this definition:
```java
// Return common api versions or empty if there are none
public Optional<ApiVersion> intersect(ApiVersion other);
```
Then we can let the caller decide how to handle this case.